{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ML Model Based on cross validation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Ok so we will be using the cross validation to train the model and then we will be using the model to predict the values of the test data. \n",
    "- We will use cross validation to find which algorithm is giving the best output and then use that as our ML model algorithm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What are the steps we will be taking to build the model?\n",
    "- We will first load the data and then do some basic data analysis\n",
    "- We will then do some data preprocessing\n",
    "- We will then use cross validation to find the best algorithm\n",
    "- We will then use the best algorithm to predict the values of the test data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The problem trying to be solved is Single-Label Multi-Class classification.\n",
    "\n",
    "We will be using the following algorithms and then use cross validation to find the best algorithm and then perform the evaluation of the model.\n",
    "\n",
    "\n",
    "Several machine learning algorithms can handle multi-class classification natively, meaning they are designed to work directly with multiple classes without needing additional strategies like One-vs-Rest (OvR) or One-vs-One (OvO). Some examples of algorithms that can handle multi-class classification include:\n",
    "- K-Nearest Neighbors (KNN) \n",
    "- decision tree classifier \n",
    "- random forest classifier \n",
    "- Gradieent Boosting Classifier \n",
    "- Naive Bayes: GaussianNB\n",
    "- lasso Regression (L1 Regularization, comes with Logistic Regression)\n",
    "- Ridge Regression (L2 Regularization, comes with Logistic Regression) \n",
    "\n",
    "Models like Support Vector Machines (SVM) and Logistic Regression are designed for binary classification and do not natively support multi-class classification. These algorithms can be used for multi-class classification by using a strategy like One-vs-Rest (OvR) or One-vs-One (OvO)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Methodology\n",
    "- Data Preprocessing\n",
    "- Data Analysis\n",
    "- Data Splitting: For Cross Validation\n",
    "- Model Building\n",
    "- Testing all models and their Cross Validation Scores\n",
    "- Finding the best model\n",
    "- Data splitting in the traditional way to train the best model\n",
    "- Model Evaluation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Importing libraries. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd #Pandas is for Data processing. \n",
    "import numpy as np #Numpy is for numerical calculations.\n",
    "import matplotlib.pyplot as plt #plotting and visually understanding the data.\n",
    "from sklearn.preprocessing import LabelEncoder, OneHotEncoder #Encoding the data depending on the data type.\n",
    "from sklearn.model_selection import train_test_split #split the data into testing and training set\n",
    "from sklearn.model_selection import cross_val_score, KFold #cross validation\n",
    "from sklearn.preprocessing import StandardScaler, MinMaxScaler, OneHotEncoder, PolynomialFeatures\n",
    "from sklearn.pipeline import Pipeline\n",
    "from sklearn.multiclass import OneVsRestClassifier\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.linear_model import LogisticRegression, Lasso, Ridge, ElasticNet\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.ensemble import GradientBoostingClassifier\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.naive_bayes import GaussianNB\n",
    "from sklearn.neural_network import MLPClassifier\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import accuracy_score, classification_report, roc_auc_score, roc_curve, auc, RocCurveDisplay\n",
    "from sklearn.preprocessing import label_binarize\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### data preprocessing and analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\satya\\AppData\\Local\\Temp\\ipykernel_30308\\2402356703.py:1: DtypeWarning: Columns (2,5,9,20,78,79,147,148,149,150,151,152,153,154,155,156,165,166,168) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  main_df = pd.read_csv(\"abundance.csv\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>sampleID</th>\n",
       "      <th>subjectID</th>\n",
       "      <th>bodysite</th>\n",
       "      <th>disease</th>\n",
       "      <th>age</th>\n",
       "      <th>gender</th>\n",
       "      <th>country</th>\n",
       "      <th>sequencing_technology</th>\n",
       "      <th>pubmedid</th>\n",
       "      <th>...</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Enterococcus|s__Enterococcus_gilvus|t__Enterococcus_gilvus_unclassified</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis|t__GCF_000415925</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis|t__GCF_000215085</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__GCF_000209855</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1|t__GCF_000478965</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WT2D</td>\n",
       "      <td>S367</td>\n",
       "      <td>s367</td>\n",
       "      <td>stool</td>\n",
       "      <td>impaired_glucose_tolerance</td>\n",
       "      <td>70.91</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yugoslavia</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>23719380</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WT2D</td>\n",
       "      <td>S604</td>\n",
       "      <td>s604</td>\n",
       "      <td>stool</td>\n",
       "      <td>t2d</td>\n",
       "      <td>71.63</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yugoslavia</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>23719380</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hmp</td>\n",
       "      <td>SRS011061</td>\n",
       "      <td>158458797</td>\n",
       "      <td>stool</td>\n",
       "      <td>n</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>usa</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>22699609</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hmp</td>\n",
       "      <td>SRS011084</td>\n",
       "      <td>158479027</td>\n",
       "      <td>stool</td>\n",
       "      <td>n</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>usa</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>22699609</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hmp</td>\n",
       "      <td>SRS011086</td>\n",
       "      <td>158458797</td>\n",
       "      <td>tongue_dorsum</td>\n",
       "      <td>n</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>usa</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>22699609</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3605</th>\n",
       "      <td>t2dmeta_long</td>\n",
       "      <td>T2D-034</td>\n",
       "      <td>-</td>\n",
       "      <td>stool</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>23023125</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3606</th>\n",
       "      <td>t2dmeta_long</td>\n",
       "      <td>T2D-035</td>\n",
       "      <td>-</td>\n",
       "      <td>stool</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>23023125</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3607</th>\n",
       "      <td>t2dmeta_long</td>\n",
       "      <td>T2D-037</td>\n",
       "      <td>-</td>\n",
       "      <td>stool</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>23023125</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3608</th>\n",
       "      <td>t2dmeta_long</td>\n",
       "      <td>T2D-038</td>\n",
       "      <td>-</td>\n",
       "      <td>stool</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>23023125</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3609</th>\n",
       "      <td>t2dmeta_long</td>\n",
       "      <td>T2D-061</td>\n",
       "      <td>-</td>\n",
       "      <td>stool</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>23023125</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3610 rows × 3513 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      dataset_name   sampleID  subjectID       bodysite  \\\n",
       "0             WT2D       S367       s367          stool   \n",
       "1             WT2D       S604       s604          stool   \n",
       "2              hmp  SRS011061  158458797          stool   \n",
       "3              hmp  SRS011084  158479027          stool   \n",
       "4              hmp  SRS011086  158458797  tongue_dorsum   \n",
       "...            ...        ...        ...            ...   \n",
       "3605  t2dmeta_long    T2D-034          -          stool   \n",
       "3606  t2dmeta_long    T2D-035          -          stool   \n",
       "3607  t2dmeta_long    T2D-037          -          stool   \n",
       "3608  t2dmeta_long    T2D-038          -          stool   \n",
       "3609  t2dmeta_long    T2D-061          -          stool   \n",
       "\n",
       "                         disease    age  gender     country  \\\n",
       "0     impaired_glucose_tolerance  70.91     NaN  yugoslavia   \n",
       "1                            t2d  71.63     NaN  yugoslavia   \n",
       "2                              n    NaN  female         usa   \n",
       "3                              n    NaN    male         usa   \n",
       "4                              n    NaN  female         usa   \n",
       "...                          ...    ...     ...         ...   \n",
       "3605                           -      -       -           -   \n",
       "3606                           -      -       -           -   \n",
       "3607                           -      -       -           -   \n",
       "3608                           -      -       -           -   \n",
       "3609                           -      -       -           -   \n",
       "\n",
       "     sequencing_technology  pubmedid  ...  \\\n",
       "0                 Illumina  23719380  ...   \n",
       "1                 Illumina  23719380  ...   \n",
       "2                 Illumina  22699609  ...   \n",
       "3                 Illumina  22699609  ...   \n",
       "4                 Illumina  22699609  ...   \n",
       "...                    ...       ...  ...   \n",
       "3605              Illumina  23023125  ...   \n",
       "3606              Illumina  23023125  ...   \n",
       "3607              Illumina  23023125  ...   \n",
       "3608              Illumina  23023125  ...   \n",
       "3609              Illumina  23023125  ...   \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Enterococcus|s__Enterococcus_gilvus|t__Enterococcus_gilvus_unclassified  \\\n",
       "0                                                   0.0                                                                                                      \n",
       "1                                                   0.0                                                                                                      \n",
       "2                                                   0.0                                                                                                      \n",
       "3                                                   0.0                                                                                                      \n",
       "4                                                   0.0                                                                                                      \n",
       "...                                                 ...                                                                                                      \n",
       "3605                                                0.0                                                                                                      \n",
       "3606                                                0.0                                                                                                      \n",
       "3607                                                0.0                                                                                                      \n",
       "3608                                                0.0                                                                                                      \n",
       "3609                                                0.0                                                                                                      \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis  \\\n",
       "0                                                   0.0                                                                         \n",
       "1                                                   0.0                                                                         \n",
       "2                                                   0.0                                                                         \n",
       "3                                                   0.0                                                                         \n",
       "4                                                   0.0                                                                         \n",
       "...                                                 ...                                                                         \n",
       "3605                                                0.0                                                                         \n",
       "3606                                                0.0                                                                         \n",
       "3607                                                0.0                                                                         \n",
       "3608                                                0.0                                                                         \n",
       "3609                                                0.0                                                                         \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis|t__GCF_000415925  \\\n",
       "0                                                   0.0                                                                                          \n",
       "1                                                   0.0                                                                                          \n",
       "2                                                   0.0                                                                                          \n",
       "3                                                   0.0                                                                                          \n",
       "4                                                   0.0                                                                                          \n",
       "...                                                 ...                                                                                          \n",
       "3605                                                0.0                                                                                          \n",
       "3606                                                0.0                                                                                          \n",
       "3607                                                0.0                                                                                          \n",
       "3608                                                0.0                                                                                          \n",
       "3609                                                0.0                                                                                          \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae  \\\n",
       "0                                                   0.0                           \n",
       "1                                                   0.0                           \n",
       "2                                                   0.0                           \n",
       "3                                                   0.0                           \n",
       "4                                                   0.0                           \n",
       "...                                                 ...                           \n",
       "3605                                                0.0                           \n",
       "3606                                                0.0                           \n",
       "3607                                                0.0                           \n",
       "3608                                                0.0                           \n",
       "3609                                                0.0                           \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum  \\\n",
       "0                                                   0.0                                               \n",
       "1                                                   0.0                                               \n",
       "2                                                   0.0                                               \n",
       "3                                                   0.0                                               \n",
       "4                                                   0.0                                               \n",
       "...                                                 ...                                               \n",
       "3605                                                0.0                                               \n",
       "3606                                                0.0                                               \n",
       "3607                                                0.0                                               \n",
       "3608                                                0.0                                               \n",
       "3609                                                0.0                                               \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis  \\\n",
       "0                                                   0.0                                                                           \n",
       "1                                                   0.0                                                                           \n",
       "2                                                   0.0                                                                           \n",
       "3                                                   0.0                                                                           \n",
       "4                                                   0.0                                                                           \n",
       "...                                                 ...                                                                           \n",
       "3605                                                0.0                                                                           \n",
       "3606                                                0.0                                                                           \n",
       "3607                                                0.0                                                                           \n",
       "3608                                                0.0                                                                           \n",
       "3609                                                0.0                                                                           \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis|t__GCF_000215085  \\\n",
       "0                                                   0.0                                                                                            \n",
       "1                                                   0.0                                                                                            \n",
       "2                                                   0.0                                                                                            \n",
       "3                                                   0.0                                                                                            \n",
       "4                                                   0.0                                                                                            \n",
       "...                                                 ...                                                                                            \n",
       "3605                                                0.0                                                                                            \n",
       "3606                                                0.0                                                                                            \n",
       "3607                                                0.0                                                                                            \n",
       "3608                                                0.0                                                                                            \n",
       "3609                                                0.0                                                                                            \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__GCF_000209855  \\\n",
       "0                                                   0.0                                                                                                 \n",
       "1                                                   0.0                                                                                                 \n",
       "2                                                   0.0                                                                                                 \n",
       "3                                                   0.0                                                                                                 \n",
       "4                                                   0.0                                                                                                 \n",
       "...                                                 ...                                                                                                 \n",
       "3605                                                0.0                                                                                                 \n",
       "3606                                                0.0                                                                                                 \n",
       "3607                                                0.0                                                                                                 \n",
       "3608                                                0.0                                                                                                 \n",
       "3609                                                0.0                                                                                                 \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1  \\\n",
       "0                                                   0.0                                                                           \n",
       "1                                                   0.0                                                                           \n",
       "2                                                   0.0                                                                           \n",
       "3                                                   0.0                                                                           \n",
       "4                                                   0.0                                                                           \n",
       "...                                                 ...                                                                           \n",
       "3605                                                0.0                                                                           \n",
       "3606                                                0.0                                                                           \n",
       "3607                                                0.0                                                                           \n",
       "3608                                                0.0                                                                           \n",
       "3609                                                0.0                                                                           \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1|t__GCF_000478965  \n",
       "0                                                   0.0                                                                                           \n",
       "1                                                   0.0                                                                                           \n",
       "2                                                   0.0                                                                                           \n",
       "3                                                   0.0                                                                                           \n",
       "4                                                   0.0                                                                                           \n",
       "...                                                 ...                                                                                           \n",
       "3605                                                0.0                                                                                           \n",
       "3606                                                0.0                                                                                           \n",
       "3607                                                0.0                                                                                           \n",
       "3608                                                0.0                                                                                           \n",
       "3609                                                0.0                                                                                           \n",
       "\n",
       "[3610 rows x 3513 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "main_df = pd.read_csv(\"abundance.csv\")\n",
    "main_df = main_df.replace('nd', np.nan)\n",
    "\n",
    "display(main_df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate and print missing values in the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing value percentages:\n",
      "dataset_name                                                                                                                                         0.000000\n",
      "sampleID                                                                                                                                             0.000000\n",
      "subjectID                                                                                                                                            0.000000\n",
      "bodysite                                                                                                                                            12.686981\n",
      "disease                                                                                                                                             13.157895\n",
      "                                                                                                                                                      ...    \n",
      "k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis                           0.000000\n",
      "k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis|t__GCF_000215085          0.000000\n",
      "k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__GCF_000209855     0.000000\n",
      "k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1                           0.000000\n",
      "k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1|t__GCF_000478965          0.000000\n",
      "Length: 3513, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "missing_percentages = main_df.isnull().mean() * 100\n",
    "print(\"Missing value percentages:\")\n",
    "print(missing_percentages)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "due to no biological relevancy we are dropping these:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bodysite</th>\n",
       "      <th>disease</th>\n",
       "      <th>age</th>\n",
       "      <th>gender</th>\n",
       "      <th>country</th>\n",
       "      <th>bmi</th>\n",
       "      <th>infant_gender</th>\n",
       "      <th>delivery_mode</th>\n",
       "      <th>infant_ethnicity</th>\n",
       "      <th>infant_race</th>\n",
       "      <th>...</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Enterococcus|s__Enterococcus_gilvus|t__Enterococcus_gilvus_unclassified</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis|t__GCF_000415925</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis|t__GCF_000215085</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__GCF_000209855</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1|t__GCF_000478965</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>stool</td>\n",
       "      <td>impaired_glucose_tolerance</td>\n",
       "      <td>70.91</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yugoslavia</td>\n",
       "      <td>32.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>stool</td>\n",
       "      <td>t2d</td>\n",
       "      <td>71.63</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yugoslavia</td>\n",
       "      <td>28.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>stool</td>\n",
       "      <td>n</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>usa</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>stool</td>\n",
       "      <td>n</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>usa</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tongue_dorsum</td>\n",
       "      <td>n</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>usa</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3605</th>\n",
       "      <td>stool</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3606</th>\n",
       "      <td>stool</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3607</th>\n",
       "      <td>stool</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3608</th>\n",
       "      <td>stool</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3609</th>\n",
       "      <td>stool</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3610 rows × 3473 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           bodysite                     disease    age  gender     country  \\\n",
       "0             stool  impaired_glucose_tolerance  70.91     NaN  yugoslavia   \n",
       "1             stool                         t2d  71.63     NaN  yugoslavia   \n",
       "2             stool                           n    NaN  female         usa   \n",
       "3             stool                           n    NaN    male         usa   \n",
       "4     tongue_dorsum                           n    NaN  female         usa   \n",
       "...             ...                         ...    ...     ...         ...   \n",
       "3605          stool                           -      -       -           -   \n",
       "3606          stool                           -      -       -           -   \n",
       "3607          stool                           -      -       -           -   \n",
       "3608          stool                           -      -       -           -   \n",
       "3609          stool                           -      -       -           -   \n",
       "\n",
       "       bmi infant_gender delivery_mode infant_ethnicity infant_race  ...  \\\n",
       "0     32.3           NaN           NaN              NaN         NaN  ...   \n",
       "1     28.9           NaN           NaN              NaN         NaN  ...   \n",
       "2      NaN           NaN           NaN              NaN         NaN  ...   \n",
       "3      NaN           NaN           NaN              NaN         NaN  ...   \n",
       "4      NaN           NaN           NaN              NaN         NaN  ...   \n",
       "...    ...           ...           ...              ...         ...  ...   \n",
       "3605     -           NaN           NaN              NaN         NaN  ...   \n",
       "3606     -           NaN           NaN              NaN         NaN  ...   \n",
       "3607     -           NaN           NaN              NaN         NaN  ...   \n",
       "3608     -           NaN           NaN              NaN         NaN  ...   \n",
       "3609     -           NaN           NaN              NaN         NaN  ...   \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Enterococcus|s__Enterococcus_gilvus|t__Enterococcus_gilvus_unclassified  \\\n",
       "0                                                   0.0                                                                                                      \n",
       "1                                                   0.0                                                                                                      \n",
       "2                                                   0.0                                                                                                      \n",
       "3                                                   0.0                                                                                                      \n",
       "4                                                   0.0                                                                                                      \n",
       "...                                                 ...                                                                                                      \n",
       "3605                                                0.0                                                                                                      \n",
       "3606                                                0.0                                                                                                      \n",
       "3607                                                0.0                                                                                                      \n",
       "3608                                                0.0                                                                                                      \n",
       "3609                                                0.0                                                                                                      \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis  \\\n",
       "0                                                   0.0                                                                         \n",
       "1                                                   0.0                                                                         \n",
       "2                                                   0.0                                                                         \n",
       "3                                                   0.0                                                                         \n",
       "4                                                   0.0                                                                         \n",
       "...                                                 ...                                                                         \n",
       "3605                                                0.0                                                                         \n",
       "3606                                                0.0                                                                         \n",
       "3607                                                0.0                                                                         \n",
       "3608                                                0.0                                                                         \n",
       "3609                                                0.0                                                                         \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis|t__GCF_000415925  \\\n",
       "0                                                   0.0                                                                                          \n",
       "1                                                   0.0                                                                                          \n",
       "2                                                   0.0                                                                                          \n",
       "3                                                   0.0                                                                                          \n",
       "4                                                   0.0                                                                                          \n",
       "...                                                 ...                                                                                          \n",
       "3605                                                0.0                                                                                          \n",
       "3606                                                0.0                                                                                          \n",
       "3607                                                0.0                                                                                          \n",
       "3608                                                0.0                                                                                          \n",
       "3609                                                0.0                                                                                          \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae  \\\n",
       "0                                                   0.0                           \n",
       "1                                                   0.0                           \n",
       "2                                                   0.0                           \n",
       "3                                                   0.0                           \n",
       "4                                                   0.0                           \n",
       "...                                                 ...                           \n",
       "3605                                                0.0                           \n",
       "3606                                                0.0                           \n",
       "3607                                                0.0                           \n",
       "3608                                                0.0                           \n",
       "3609                                                0.0                           \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum  \\\n",
       "0                                                   0.0                                               \n",
       "1                                                   0.0                                               \n",
       "2                                                   0.0                                               \n",
       "3                                                   0.0                                               \n",
       "4                                                   0.0                                               \n",
       "...                                                 ...                                               \n",
       "3605                                                0.0                                               \n",
       "3606                                                0.0                                               \n",
       "3607                                                0.0                                               \n",
       "3608                                                0.0                                               \n",
       "3609                                                0.0                                               \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis  \\\n",
       "0                                                   0.0                                                                           \n",
       "1                                                   0.0                                                                           \n",
       "2                                                   0.0                                                                           \n",
       "3                                                   0.0                                                                           \n",
       "4                                                   0.0                                                                           \n",
       "...                                                 ...                                                                           \n",
       "3605                                                0.0                                                                           \n",
       "3606                                                0.0                                                                           \n",
       "3607                                                0.0                                                                           \n",
       "3608                                                0.0                                                                           \n",
       "3609                                                0.0                                                                           \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis|t__GCF_000215085  \\\n",
       "0                                                   0.0                                                                                            \n",
       "1                                                   0.0                                                                                            \n",
       "2                                                   0.0                                                                                            \n",
       "3                                                   0.0                                                                                            \n",
       "4                                                   0.0                                                                                            \n",
       "...                                                 ...                                                                                            \n",
       "3605                                                0.0                                                                                            \n",
       "3606                                                0.0                                                                                            \n",
       "3607                                                0.0                                                                                            \n",
       "3608                                                0.0                                                                                            \n",
       "3609                                                0.0                                                                                            \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__GCF_000209855  \\\n",
       "0                                                   0.0                                                                                                 \n",
       "1                                                   0.0                                                                                                 \n",
       "2                                                   0.0                                                                                                 \n",
       "3                                                   0.0                                                                                                 \n",
       "4                                                   0.0                                                                                                 \n",
       "...                                                 ...                                                                                                 \n",
       "3605                                                0.0                                                                                                 \n",
       "3606                                                0.0                                                                                                 \n",
       "3607                                                0.0                                                                                                 \n",
       "3608                                                0.0                                                                                                 \n",
       "3609                                                0.0                                                                                                 \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1  \\\n",
       "0                                                   0.0                                                                           \n",
       "1                                                   0.0                                                                           \n",
       "2                                                   0.0                                                                           \n",
       "3                                                   0.0                                                                           \n",
       "4                                                   0.0                                                                           \n",
       "...                                                 ...                                                                           \n",
       "3605                                                0.0                                                                           \n",
       "3606                                                0.0                                                                           \n",
       "3607                                                0.0                                                                           \n",
       "3608                                                0.0                                                                           \n",
       "3609                                                0.0                                                                           \n",
       "\n",
       "     k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1|t__GCF_000478965  \n",
       "0                                                   0.0                                                                                           \n",
       "1                                                   0.0                                                                                           \n",
       "2                                                   0.0                                                                                           \n",
       "3                                                   0.0                                                                                           \n",
       "4                                                   0.0                                                                                           \n",
       "...                                                 ...                                                                                           \n",
       "3605                                                0.0                                                                                           \n",
       "3606                                                0.0                                                                                           \n",
       "3607                                                0.0                                                                                           \n",
       "3608                                                0.0                                                                                           \n",
       "3609                                                0.0                                                                                           \n",
       "\n",
       "[3610 rows x 3473 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "# Create a copy of main_df before dropping columns\n",
    "main_df_cleaned = main_df.copy()\n",
    "\n",
    "# Drop columns from main_df_cleaned instead of main_df\n",
    "columns_to_drop = ['dataset_name', 'sampleID', 'subjectID', 'sequencing_technology', 'pubmedid', 'camp', \n",
    "                   'collectionweek', 'samplecollectionwindow', 'paired_end_insert_size', 'read_length', \n",
    "                   'total_reads', 'matched_reads', 'uniquely_matching_reads', 'uniquely_matched_reads', \n",
    "                   'gene_number', 'gene_number_for_11m_uniquely_matched_reads', 'hitchip_probe_number', \n",
    "                   'gene_count_class', 'hitchip_probe_class', '#SampleID', 'rna_sampleid', \n",
    "                   'postnatal_antimicrobial_use', 'infant_gestation_weeks', 'cohort', 'less_than_29weeks', \n",
    "                   'sample_collection_days', 'gut_sample_id_ncbipublic', 'gut_sample_id_corrected', \n",
    "                   'projectid', 'flowcell', 'comment', 'mlst_project', 'mlst_ec', 'st_ec', 'mlst_kp', \n",
    "                   'st_kp', 'extractionprotocolid', 'site_id_cincinnati', 'ascites', 'classification']\n",
    "\n",
    "main_df_cleaned.drop(columns=columns_to_drop, inplace=True, errors='ignore')\n",
    "display(main_df_cleaned)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, coming to data cleaning, we are using mean imputation for the numerical columns only and forward fill for the categorical columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\satya\\AppData\\Local\\Temp\\ipykernel_30308\\2386306714.py:11: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.\n",
      "  df_cleaned[categorical_columns] = df_cleaned[categorical_columns].fillna(method='ffill')\n"
     ]
    }
   ],
   "source": [
    "from sklearn.impute import SimpleImputer\n",
    "\n",
    "# Assuming main_df_cleaned is the cleaned dataframe from previous steps\n",
    "df_cleaned = main_df_cleaned.copy()\n",
    "\n",
    "# Define categorical and numerical columns\n",
    "categorical_columns = df_cleaned.select_dtypes(include=['object']).columns\n",
    "numerical_columns = df_cleaned.select_dtypes(include=[np.number]).columns\n",
    "\n",
    "# For categorical columns, use forward fill\n",
    "df_cleaned[categorical_columns] = df_cleaned[categorical_columns].fillna(method='ffill')\n",
    "\n",
    "# For numerical columns, use mean imputation\n",
    "numerical_imputer = SimpleImputer(strategy='mean')\n",
    "df_cleaned[numerical_columns] = numerical_imputer.fit_transform(df_cleaned[numerical_columns])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lets once view and export the dataset now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "export_csv = df_cleaned.to_csv(r'abundance_cleaned.csv', index = None, header=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Columns in the current dataset \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['bodysite', 'disease', 'age', 'gender', 'country', 'bmi', 'infant_gender', 'delivery_mode', 'infant_ethnicity', 'infant_race', 'birth_year', 'necrotizing_enterocolitis', 'dol_firstnecors', 'sepsis', 'infant_birthweight_kg', 'infant_birth_length_cm', 'maternal_abx_given', 'postnatal_abx_2window', 'perinatal_antimicrobial_use', 'death', 'nursing_status', 'maternal_age_at_delivery_years', 'mat_age_bin', 'parity', 'infant_fut2', 'mat_fut2', 'lowh', 'lowlea', 'highslea', 'married', 'gravida', 'preeclampsia', 'mult_birth', 'hypertension', 'hypertension_prepreg', 'chorioamnionitis', 'primipar', 'daysonform3daysprior', 'days_on_abx_14', 'visit_number', 'snprnt', 'wmsphase', 'first', 'repeat', 'stooltexture', 'daysafteronset', 'hus', 'stec_count', 'shigatoxin2elisa', 'readsmillions', 'nonhuman', 'stec_coverage', 'stxab_detected', 'stx_ratio', 'typingdata', 'c_difficile_frequency', 'ibd', 'sampling_day', 'known_consumers_of_a_defined_fermented_milk_product_(dfmp)', 'mgs_richness', 'mgs_profile_matched_sample_pairs', 'ethnicity', 'sample_name', 'index_2_primer', 'index_1_primer', 'sample_name_repeat1', 'extraction_conc', 'sampleid_with_barcode', 'initial_nreads', 'quality_control', 'screensensloc_hg19', 'screensensloc_phix174', 'screenhiv', 'percent_initial_reads', 'percent_qc_reads', 'tipo_prelievo', 'prima_dopo_cura', 'codice_provetta', 'data', 'prelievo_efemalefemaleettuato_presso', 'patient_number', 'pasi', 'bsa', 'age_ofemale_onset', 'arthritis', 'type', 'antibiotic_usage', 'stage', 'cirrhotic', 'hbv_related', 'alcohol_related', 'other_causes_related', 'inr', 'crea', 'alb', 'tb', 'pt', 'he', 'ctp', 'meld', 'antivirus', '?-blocker', 'designation', 'age_range', 'site_symmetry', 'affected/unaffected', 'method', 'site_characteristic', 'total_intital_reads', 'estimated_median_insert_size', 'reported_as_failed_qc', 'uniquely_align_to_human', 'non_uniquely_align_to_human_with_0_2_mismatches', 'reads_removed_because_of_read_pair_trimming_discrepancy', 'too_short_after_quality_trimming(<50bp)', 'final_#_paired_reads', 'final_#_singletons', 'total_final_reads', '%_reads_kept_by_mean_slope_of_last_5_points_(coverage)', 'population', 'bmi_class', '16s_rrna', 'shotgun_metagenome', 'read_depth_16s', 'paired', 'single', 'height', 'weight', 'diabetic', 'fbg', 'sbp', 'dbp', 'fins', 'fcp', 'hbalc', 'tg', 'tcho', 'hdl', 'ldl', 'kit', 'sample_group', 'illumina_run', 'gad-antibodies', 'whr', 'wc', 'cholesterol', 'triglycerides', 'creatinine', 'y-gt', 'fasting_glucose', 'fasting_insulin', 'hba1c', 'adiponectin', 'leptin', 'glp-1', 'fgf-19', 'hscrp', 'c-peptide', 'tnfa', 'il-1', 'cd163', 'statins', 'insulin', 'oral_anti-diabetic_medication', 'years_in_sweden', 'tnm_stage', 'ajcc_stage', 'localization', 'fobt', 'wif-1_gene_methylation_test', 'group']\n"
     ]
    }
   ],
   "source": [
    "    column_name = df_cleaned.select_dtypes(exclude=[np.number]).columns.tolist()\n",
    "    print(column_name)\n",
    "    column_name = df_cleaned.select_dtypes(exclude=[np.number]).columns.tolist()\n",
    "    print(column_name)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data splitting (X AND Y)\n",
    "why are we not encoding before this? reason is that you are not supposed to encode the target variable.\n",
    "\n",
    "We are now considering everything else except the target variable as the features and the target variable as the target."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        bodysite    age  gender     country   bmi infant_gender delivery_mode  \\\n",
      "0          stool  70.91     NaN  yugoslavia  32.3           NaN           NaN   \n",
      "1          stool  71.63     NaN  yugoslavia  28.9           NaN           NaN   \n",
      "2          stool  71.63  female         usa  28.9           NaN           NaN   \n",
      "3          stool  71.63    male         usa  28.9           NaN           NaN   \n",
      "4  tongue_dorsum  71.63  female         usa  28.9           NaN           NaN   \n",
      "\n",
      "  infant_ethnicity infant_race birth_year  ...  \\\n",
      "0              NaN         NaN        NaN  ...   \n",
      "1              NaN         NaN        NaN  ...   \n",
      "2              NaN         NaN        NaN  ...   \n",
      "3              NaN         NaN        NaN  ...   \n",
      "4              NaN         NaN        NaN  ...   \n",
      "\n",
      "  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Enterococcus|s__Enterococcus_gilvus|t__Enterococcus_gilvus_unclassified  \\\n",
      "0                                                0.0                                                                                                      \n",
      "1                                                0.0                                                                                                      \n",
      "2                                                0.0                                                                                                      \n",
      "3                                                0.0                                                                                                      \n",
      "4                                                0.0                                                                                                      \n",
      "\n",
      "  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis  \\\n",
      "0                                                0.0                                                                         \n",
      "1                                                0.0                                                                         \n",
      "2                                                0.0                                                                         \n",
      "3                                                0.0                                                                         \n",
      "4                                                0.0                                                                         \n",
      "\n",
      "  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis|t__GCF_000415925  \\\n",
      "0                                                0.0                                                                                          \n",
      "1                                                0.0                                                                                          \n",
      "2                                                0.0                                                                                          \n",
      "3                                                0.0                                                                                          \n",
      "4                                                0.0                                                                                          \n",
      "\n",
      "  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae  \\\n",
      "0                                                0.0                           \n",
      "1                                                0.0                           \n",
      "2                                                0.0                           \n",
      "3                                                0.0                           \n",
      "4                                                0.0                           \n",
      "\n",
      "  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum  \\\n",
      "0                                                0.0                                               \n",
      "1                                                0.0                                               \n",
      "2                                                0.0                                               \n",
      "3                                                0.0                                               \n",
      "4                                                0.0                                               \n",
      "\n",
      "  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis  \\\n",
      "0                                                0.0                                                                           \n",
      "1                                                0.0                                                                           \n",
      "2                                                0.0                                                                           \n",
      "3                                                0.0                                                                           \n",
      "4                                                0.0                                                                           \n",
      "\n",
      "  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis|t__GCF_000215085  \\\n",
      "0                                                0.0                                                                                            \n",
      "1                                                0.0                                                                                            \n",
      "2                                                0.0                                                                                            \n",
      "3                                                0.0                                                                                            \n",
      "4                                                0.0                                                                                            \n",
      "\n",
      "  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__GCF_000209855  \\\n",
      "0                                                0.0                                                                                                 \n",
      "1                                                0.0                                                                                                 \n",
      "2                                                0.0                                                                                                 \n",
      "3                                                0.0                                                                                                 \n",
      "4                                                0.0                                                                                                 \n",
      "\n",
      "  k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1  \\\n",
      "0                                                0.0                                                                           \n",
      "1                                                0.0                                                                           \n",
      "2                                                0.0                                                                           \n",
      "3                                                0.0                                                                           \n",
      "4                                                0.0                                                                           \n",
      "\n",
      "  k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1|t__GCF_000478965  \n",
      "0                                                0.0                                                                                           \n",
      "1                                                0.0                                                                                           \n",
      "2                                                0.0                                                                                           \n",
      "3                                                0.0                                                                                           \n",
      "4                                                0.0                                                                                           \n",
      "\n",
      "[5 rows x 3472 columns]\n",
      "0    impaired_glucose_tolerance\n",
      "1                           t2d\n",
      "2                             n\n",
      "3                             n\n",
      "4                             n\n",
      "Name: disease, dtype: object\n"
     ]
    }
   ],
   "source": [
    "# Note that disease column has already been dropped when creating X\n",
    "Input_Features = [col for col in df_cleaned.columns if col != 'disease']\n",
    "X = df_cleaned[Input_Features]\n",
    "Y = df_cleaned['disease']\n",
    "\n",
    "print(X.head())\n",
    "\n",
    "print(Y.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now lets do some encoding for the categorical columns\n",
    "now, we will do two types of encoding i.e. label encoding and one hot encoding.\n",
    "\n",
    "We are doing this because encoding has an impact in complexity of the model which will impact things like bias and variance and also the accuracy of the model\n",
    "\n",
    "label encoding assigns unique value to the categories and one hot encoding creates a new column for each category and assigns a binary value to it\n",
    "\n",
    "- Label encoding works better for algorithms that DO NOT take ordinality into consideration. One hot encoding works better for algorithms that take ordinality into consideration\n",
    "\n",
    "- label encoding works better for decision trees and random forests and one hot encoding works better for linear regression and logistic regression\n",
    "\n",
    "- for distance dependent algorithms like KNN, SVM, Neural Networks one hot encoding is preferred\n",
    "\n",
    "we will bedoing encoding only to X axis\n",
    "algrmths like decision trees, random forests, naive bayes dont really need encoding but we will apply label encoding for them"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "which encoding to use for which algorithm:\n",
    "- Decision Trees and Random Forests: Label Encoding\n",
    "- Naive Bayes: Label Encoding\n",
    "- KNN, SVM, Logistic regression: One Hot Encoding\n",
    "- lasso and ridge regression: One Hot Encoding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find numeric and non numeric columns once\n",
    "numeric_columns = X.select_dtypes(include=[np.number]).columns\n",
    "non_numeric_columns = X.select_dtypes(include=[object]).columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CONVERT ALL CATEGORICAL DATA TO STRINGS\n",
    "X_for_encoding = X.copy()\n",
    "for column in non_numeric_columns:\n",
    "\tX_for_encoding[column] = X_for_encoding[column].astype(str)\n",
    "\n",
    "label_encoders = {}\n",
    "Label_X = X_for_encoding.copy()\n",
    "for column in non_numeric_columns:\n",
    "\tlabel_encoders[column] = LabelEncoder()\n",
    "\tLabel_X[column] = label_encoders[column].fit_transform(X_for_encoding[column])\n",
    "\n",
    "onehot_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')\n",
    "Onehot_X = onehot_encoder.fit_transform(X_for_encoding[non_numeric_columns])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "now coming back to scaling, we are using MinMaXscalar for the numerical columns only."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "scaler_minmax = MinMaxScaler()\n",
    "df_cleaned[numerical_columns] = scaler_minmax.fit_transform(df_cleaned[numerical_columns])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cross Validation technique.\n",
    "Cross-Validation is a technique used to assess the performance of a machine learning model more reliably by separating the data into 'folds' then and using different folds for training and testing. \n",
    "\n",
    "this has many advantages including: handling bias and variance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "kf = KFold(n_splits=10, shuffle=True, random_state=42)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Models to Evaluate:\n",
    "- SVC (one vs rest)\n",
    "- Logistic Regression (one vs rest)\n",
    "- K-Nearest Neighbors (KNN)\n",
    "- Decision Tree Classifier\n",
    "- Random Forest Classifier\n",
    "- MLP Classifier\n",
    "- Naive Bayes: GaussianNB\n",
    "- Regularized Regression: Lasso\n",
    "- Regularized Regression: Ridge"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lets go one by one and perform cross validation checks on each of the models."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. SVC (one vs rest)\n",
    "- SVC is a binary classifier and does not support multi-class classification natively. hence we will use OneVsRestClassifier to perform multi-class classification.\n",
    "- We will use the cross_val_score function to perform cross validation on the model.\n",
    "- for onehot_X as X because it is a distance dependent algorithm\n",
    "\n",
    "Support vector machines is "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.9146814404432135\n"
     ]
    }
   ],
   "source": [
    "SVC_OVR = SVC(decision_function_shape='ovr') # OneV Rest approc\n",
    "\n",
    "# score\n",
    "cv = KFold(n_splits=10, random_state=1, shuffle=True)\n",
    "scores = cross_val_score(SVC_OVR, Onehot_X, Y, scoring='accuracy', cv=cv, n_jobs=-1)\n",
    "# n_jobs=-1: This tells scikit-learn to use all available CPU cores.\n",
    "SVCMEAN = np.mean(scores) \n",
    "print(\"Accuracy:\", SVCMEAN)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Logistic Regression (one vs rest)\n",
    "- Logistic Regression is a binary classifier and does not support multi-class classification natively. hence we will use OneVsRestClassifier to perform multi-class classification.\n",
    "- We will use the cross_val_score function to perform cross validation on the model.\n",
    "- for onehot_X as X because it is a distance dependent algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.9570637119113574\n"
     ]
    }
   ],
   "source": [
    "LR_OVR = LogisticRegression(multi_class='ovr', solver='liblinear') \n",
    "# solver in logistic regression\n",
    "cv = KFold(n_splits=10, random_state=1, shuffle=True)\n",
    "scores = cross_val_score(LR_OVR, Onehot_X, Y, scoring='accuracy', cv=cv, n_jobs=-1)\n",
    "LRMEAN = np.mean(scores) \n",
    "print(\"Accuracy:\", LRMEAN)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. K-Nearest Neighbors (KNN)\n",
    "- KNN is a distance dependent algorithm and hence we will use one hot encoding for the categorical columns.\n",
    "- We will use the cross_val_score function to perform cross validation on the model.\n",
    "- for onehot_X as X because it is a distance dependent algorithm\n",
    "\n",
    "in knn, we take the nearest k points and then take the majority of the points to classify the point"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.9193905817174517\n"
     ]
    }
   ],
   "source": [
    "KNN = KNeighborsClassifier()\n",
    "cv = KFold(n_splits=10, random_state=1, shuffle=True)\n",
    "scores = cross_val_score(KNN, Onehot_X, Y, scoring='accuracy', cv=cv, n_jobs=-1)\n",
    "KNNMEAN = np.mean(scores) \n",
    "print(\"Accuracy:\", KNNMEAN)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# 4. Decision Tree Classifier\n",
    "- Decision Tree Classifier is a non-parametric supervised learning method used for classification and regression.\n",
    "- they are a part of ensemble methods, which are a collection of models that are trained and evaluated in parallel and then combined to make a final prediction.\n",
    "- they are the best as they can handle both numerical and categorical data and they are easy to interpret and understand. this can help in reducing complexity also.\n",
    "- and also they can help in balancing the variance bias tradeoff."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.9592797783933518\n"
     ]
    }
   ],
   "source": [
    "DT = DecisionTreeClassifier()\n",
    "cv = KFold(n_splits=10, random_state=1, shuffle=True)\n",
    "scores = cross_val_score(DT, Onehot_X, Y, scoring='accuracy', cv=cv, n_jobs=-1)\n",
    "DTMEAN = np.mean(scores)\n",
    "print(\"Accuracy:\", DTMEAN)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Random Forest Classifier\n",
    "- Random Forest Classifier is also ensemble method that uses multiple decision trees to make a prediction.\n",
    "- it is a collection of decision trees that are trained and evaluated in parallel and then combined to make a final prediction. *basically, decision trees pro max*\n",
    "- it is a good choice for multi-class classification problems because it can handle both numerical and categorical data and it is robust to overfitting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.954016620498615\n"
     ]
    }
   ],
   "source": [
    "RTC = RandomForestClassifier()\n",
    "cv = KFold(n_splits=10, random_state=1, shuffle=True)\n",
    "scores = cross_val_score(RTC, Onehot_X, Y, scoring='accuracy', cv=cv, n_jobs=-1)\n",
    "RTCMEAN = np.mean(scores)\n",
    "print(\"Accuracy:\", RTCMEAN)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Gradient Boosting Classifier\n",
    "- Gradient Boosting Classifier is a machine learning technique for regression and classification problems, which produces a prediction model in the form of an ensemble of weak prediction models, typically decision trees.\n",
    "- Gradiant boosting helps deal with loss function. it has optimisation techniques that help in reducing the loss function.\n",
    "- it is a good choice for multi-class classification problems because it can handle both numerical and categorical data and it is robust to overfitting.\n",
    "\n",
    "\n",
    "ONLY ISSUE WITH GRADIENT BOOSTING IS PERFORMANCE AND TIME ISSUE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[36], line 3\u001b[0m\n\u001b[0;32m      1\u001b[0m GBC \u001b[38;5;241m=\u001b[39m GradientBoostingClassifier()\n\u001b[0;32m      2\u001b[0m cv \u001b[38;5;241m=\u001b[39m KFold(n_splits\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m10\u001b[39m, random_state\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m, shuffle\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m----> 3\u001b[0m scores \u001b[38;5;241m=\u001b[39m \u001b[43mcross_val_score\u001b[49m\u001b[43m(\u001b[49m\u001b[43mGBC\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mOnehot_X\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mY\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mscoring\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43maccuracy\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcv\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcv\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn_jobs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m-\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[0;32m      4\u001b[0m GBCMEAN \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mmeans(cores)\n\u001b[0;32m      5\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAccuracy:\u001b[39m\u001b[38;5;124m\"\u001b[39m, GBCMEAN)\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\sklearn\\utils\\_param_validation.py:216\u001b[0m, in \u001b[0;36mvalidate_params.<locals>.decorator.<locals>.wrapper\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m    210\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m    211\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m config_context(\n\u001b[0;32m    212\u001b[0m         skip_parameter_validation\u001b[38;5;241m=\u001b[39m(\n\u001b[0;32m    213\u001b[0m             prefer_skip_nested_validation \u001b[38;5;129;01mor\u001b[39;00m global_skip_validation\n\u001b[0;32m    214\u001b[0m         )\n\u001b[0;32m    215\u001b[0m     ):\n\u001b[1;32m--> 216\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    217\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m InvalidParameterError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m    218\u001b[0m     \u001b[38;5;66;03m# When the function is just a wrapper around an estimator, we allow\u001b[39;00m\n\u001b[0;32m    219\u001b[0m     \u001b[38;5;66;03m# the function to delegate validation to the estimator, but we replace\u001b[39;00m\n\u001b[0;32m    220\u001b[0m     \u001b[38;5;66;03m# the name of the estimator by the name of the function in the error\u001b[39;00m\n\u001b[0;32m    221\u001b[0m     \u001b[38;5;66;03m# message to avoid confusion.\u001b[39;00m\n\u001b[0;32m    222\u001b[0m     msg \u001b[38;5;241m=\u001b[39m re\u001b[38;5;241m.\u001b[39msub(\n\u001b[0;32m    223\u001b[0m         \u001b[38;5;124mr\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mparameter of \u001b[39m\u001b[38;5;124m\\\u001b[39m\u001b[38;5;124mw+ must be\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[0;32m    224\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mparameter of \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfunc\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__qualname__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m must be\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[0;32m    225\u001b[0m         \u001b[38;5;28mstr\u001b[39m(e),\n\u001b[0;32m    226\u001b[0m     )\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\sklearn\\model_selection\\_validation.py:684\u001b[0m, in \u001b[0;36mcross_val_score\u001b[1;34m(estimator, X, y, groups, scoring, cv, n_jobs, verbose, params, pre_dispatch, error_score)\u001b[0m\n\u001b[0;32m    681\u001b[0m \u001b[38;5;66;03m# To ensure multimetric format is not supported\u001b[39;00m\n\u001b[0;32m    682\u001b[0m scorer \u001b[38;5;241m=\u001b[39m check_scoring(estimator, scoring\u001b[38;5;241m=\u001b[39mscoring)\n\u001b[1;32m--> 684\u001b[0m cv_results \u001b[38;5;241m=\u001b[39m \u001b[43mcross_validate\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    685\u001b[0m \u001b[43m    \u001b[49m\u001b[43mestimator\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mestimator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    686\u001b[0m \u001b[43m    \u001b[49m\u001b[43mX\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    687\u001b[0m \u001b[43m    \u001b[49m\u001b[43my\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    688\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgroups\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgroups\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    689\u001b[0m \u001b[43m    \u001b[49m\u001b[43mscoring\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m{\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mscore\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mscorer\u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    690\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcv\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcv\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    691\u001b[0m \u001b[43m    \u001b[49m\u001b[43mn_jobs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mn_jobs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    692\u001b[0m \u001b[43m    \u001b[49m\u001b[43mverbose\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mverbose\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    693\u001b[0m \u001b[43m    \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    694\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpre_dispatch\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpre_dispatch\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    695\u001b[0m \u001b[43m    \u001b[49m\u001b[43merror_score\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merror_score\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    696\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    697\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m cv_results[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtest_score\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\sklearn\\utils\\_param_validation.py:216\u001b[0m, in \u001b[0;36mvalidate_params.<locals>.decorator.<locals>.wrapper\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m    210\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m    211\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m config_context(\n\u001b[0;32m    212\u001b[0m         skip_parameter_validation\u001b[38;5;241m=\u001b[39m(\n\u001b[0;32m    213\u001b[0m             prefer_skip_nested_validation \u001b[38;5;129;01mor\u001b[39;00m global_skip_validation\n\u001b[0;32m    214\u001b[0m         )\n\u001b[0;32m    215\u001b[0m     ):\n\u001b[1;32m--> 216\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    217\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m InvalidParameterError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m    218\u001b[0m     \u001b[38;5;66;03m# When the function is just a wrapper around an estimator, we allow\u001b[39;00m\n\u001b[0;32m    219\u001b[0m     \u001b[38;5;66;03m# the function to delegate validation to the estimator, but we replace\u001b[39;00m\n\u001b[0;32m    220\u001b[0m     \u001b[38;5;66;03m# the name of the estimator by the name of the function in the error\u001b[39;00m\n\u001b[0;32m    221\u001b[0m     \u001b[38;5;66;03m# message to avoid confusion.\u001b[39;00m\n\u001b[0;32m    222\u001b[0m     msg \u001b[38;5;241m=\u001b[39m re\u001b[38;5;241m.\u001b[39msub(\n\u001b[0;32m    223\u001b[0m         \u001b[38;5;124mr\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mparameter of \u001b[39m\u001b[38;5;124m\\\u001b[39m\u001b[38;5;124mw+ must be\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[0;32m    224\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mparameter of \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfunc\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__qualname__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m must be\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[0;32m    225\u001b[0m         \u001b[38;5;28mstr\u001b[39m(e),\n\u001b[0;32m    226\u001b[0m     )\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\sklearn\\model_selection\\_validation.py:411\u001b[0m, in \u001b[0;36mcross_validate\u001b[1;34m(estimator, X, y, groups, scoring, cv, n_jobs, verbose, params, pre_dispatch, return_train_score, return_estimator, return_indices, error_score)\u001b[0m\n\u001b[0;32m    408\u001b[0m \u001b[38;5;66;03m# We clone the estimator to make sure that all the folds are\u001b[39;00m\n\u001b[0;32m    409\u001b[0m \u001b[38;5;66;03m# independent, and that it is pickle-able.\u001b[39;00m\n\u001b[0;32m    410\u001b[0m parallel \u001b[38;5;241m=\u001b[39m Parallel(n_jobs\u001b[38;5;241m=\u001b[39mn_jobs, verbose\u001b[38;5;241m=\u001b[39mverbose, pre_dispatch\u001b[38;5;241m=\u001b[39mpre_dispatch)\n\u001b[1;32m--> 411\u001b[0m results \u001b[38;5;241m=\u001b[39m \u001b[43mparallel\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    412\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdelayed\u001b[49m\u001b[43m(\u001b[49m\u001b[43m_fit_and_score\u001b[49m\u001b[43m)\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    413\u001b[0m \u001b[43m        \u001b[49m\u001b[43mclone\u001b[49m\u001b[43m(\u001b[49m\u001b[43mestimator\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    414\u001b[0m \u001b[43m        \u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    415\u001b[0m \u001b[43m        \u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    416\u001b[0m \u001b[43m        \u001b[49m\u001b[43mscorer\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mscorers\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    417\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtrain\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtrain\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    418\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtest\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtest\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    419\u001b[0m \u001b[43m        \u001b[49m\u001b[43mverbose\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mverbose\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    420\u001b[0m \u001b[43m        \u001b[49m\u001b[43mparameters\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[0;32m    421\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfit_params\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrouted_params\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mestimator\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    422\u001b[0m \u001b[43m        \u001b[49m\u001b[43mscore_params\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrouted_params\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mscorer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mscore\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    423\u001b[0m \u001b[43m        \u001b[49m\u001b[43mreturn_train_score\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreturn_train_score\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    424\u001b[0m \u001b[43m        \u001b[49m\u001b[43mreturn_times\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[0;32m    425\u001b[0m \u001b[43m        \u001b[49m\u001b[43mreturn_estimator\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreturn_estimator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    426\u001b[0m \u001b[43m        \u001b[49m\u001b[43merror_score\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merror_score\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    427\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    428\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mtrain\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtest\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mindices\u001b[49m\n\u001b[0;32m    429\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    431\u001b[0m _warn_or_raise_about_fit_failures(results, error_score)\n\u001b[0;32m    433\u001b[0m \u001b[38;5;66;03m# For callable scoring, the return type is only know after calling. If the\u001b[39;00m\n\u001b[0;32m    434\u001b[0m \u001b[38;5;66;03m# return type is a dictionary, the error scores can now be inserted with\u001b[39;00m\n\u001b[0;32m    435\u001b[0m \u001b[38;5;66;03m# the correct key.\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\sklearn\\utils\\parallel.py:77\u001b[0m, in \u001b[0;36mParallel.__call__\u001b[1;34m(self, iterable)\u001b[0m\n\u001b[0;32m     72\u001b[0m config \u001b[38;5;241m=\u001b[39m get_config()\n\u001b[0;32m     73\u001b[0m iterable_with_config \u001b[38;5;241m=\u001b[39m (\n\u001b[0;32m     74\u001b[0m     (_with_config(delayed_func, config), args, kwargs)\n\u001b[0;32m     75\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m delayed_func, args, kwargs \u001b[38;5;129;01min\u001b[39;00m iterable\n\u001b[0;32m     76\u001b[0m )\n\u001b[1;32m---> 77\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__call__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43miterable_with_config\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\joblib\\parallel.py:2007\u001b[0m, in \u001b[0;36mParallel.__call__\u001b[1;34m(self, iterable)\u001b[0m\n\u001b[0;32m   2001\u001b[0m \u001b[38;5;66;03m# The first item from the output is blank, but it makes the interpreter\u001b[39;00m\n\u001b[0;32m   2002\u001b[0m \u001b[38;5;66;03m# progress until it enters the Try/Except block of the generator and\u001b[39;00m\n\u001b[0;32m   2003\u001b[0m \u001b[38;5;66;03m# reaches the first `yield` statement. This starts the asynchronous\u001b[39;00m\n\u001b[0;32m   2004\u001b[0m \u001b[38;5;66;03m# dispatch of the tasks to the workers.\u001b[39;00m\n\u001b[0;32m   2005\u001b[0m \u001b[38;5;28mnext\u001b[39m(output)\n\u001b[1;32m-> 2007\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m output \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mreturn_generator \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;43mlist\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43moutput\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\joblib\\parallel.py:1650\u001b[0m, in \u001b[0;36mParallel._get_outputs\u001b[1;34m(self, iterator, pre_dispatch)\u001b[0m\n\u001b[0;32m   1647\u001b[0m     \u001b[38;5;28;01myield\u001b[39;00m\n\u001b[0;32m   1649\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backend\u001b[38;5;241m.\u001b[39mretrieval_context():\n\u001b[1;32m-> 1650\u001b[0m         \u001b[38;5;28;01myield from\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_retrieve()\n\u001b[0;32m   1652\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mGeneratorExit\u001b[39;00m:\n\u001b[0;32m   1653\u001b[0m     \u001b[38;5;66;03m# The generator has been garbage collected before being fully\u001b[39;00m\n\u001b[0;32m   1654\u001b[0m     \u001b[38;5;66;03m# consumed. This aborts the remaining tasks if possible and warn\u001b[39;00m\n\u001b[0;32m   1655\u001b[0m     \u001b[38;5;66;03m# the user if necessary.\u001b[39;00m\n\u001b[0;32m   1656\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_exception \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\joblib\\parallel.py:1762\u001b[0m, in \u001b[0;36mParallel._retrieve\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m   1757\u001b[0m \u001b[38;5;66;03m# If the next job is not ready for retrieval yet, we just wait for\u001b[39;00m\n\u001b[0;32m   1758\u001b[0m \u001b[38;5;66;03m# async callbacks to progress.\u001b[39;00m\n\u001b[0;32m   1759\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m ((\u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_jobs) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m\n\u001b[0;32m   1760\u001b[0m     (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_jobs[\u001b[38;5;241m0\u001b[39m]\u001b[38;5;241m.\u001b[39mget_status(\n\u001b[0;32m   1761\u001b[0m         timeout\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtimeout) \u001b[38;5;241m==\u001b[39m TASK_PENDING)):\n\u001b[1;32m-> 1762\u001b[0m     \u001b[43mtime\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msleep\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m0.01\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1763\u001b[0m     \u001b[38;5;28;01mcontinue\u001b[39;00m\n\u001b[0;32m   1765\u001b[0m \u001b[38;5;66;03m# We need to be careful: the job list can be filling up as\u001b[39;00m\n\u001b[0;32m   1766\u001b[0m \u001b[38;5;66;03m# we empty it and Python list are not thread-safe by\u001b[39;00m\n\u001b[0;32m   1767\u001b[0m \u001b[38;5;66;03m# default hence the use of the lock\u001b[39;00m\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "GBC = GradientBoostingClassifier()\n",
    "cv = KFold(n_splits=10, random_state=1, shuffle=True)\n",
    "scores = cross_val_score(GBC, Onehot_X, Y, scoring='accuracy', cv=cv, n_jobs=-1)\n",
    "GBCMEAN = np.means(scores)\n",
    "print(\"Accuracy:\", GBCMEAN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Naive Bayes: GaussianNB\n",
    "- Naive Bayes is basically uses the principles of Baye's theorem\n",
    "- in our case we will use GaussianNB because we have continuous data\n",
    "and other variants are for text based classification or count based classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.9121883656509693\n"
     ]
    }
   ],
   "source": [
    "NB = GaussianNB()\n",
    "cv = KFold(n_splits=10, random_state=1, shuffle=True)\n",
    "scores = cross_val_score(NB, Onehot_X, Y, scoring='accuracy', cv=cv, n_jobs=-1)\n",
    "NBMEAN = np.mean(scores)\n",
    "print(\"Accuracy:\", NBMEAN)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 8. MLP Classifier\n",
    "this uses neural networks to classify the data and it is a good choice as it can handle both numerical and categorical data and it is robust to overfitting.\n",
    "\n",
    "even we don't know much about neural networks but its an attempt of understnading.\n",
    "\n",
    "it is also super duper computationally intensive \n",
    "\n",
    "you need one hot encoding for this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.8360110803324099\n"
     ]
    }
   ],
   "source": [
    "MLP = MLPClassifier(hidden_layer_sizes=(10,), activation='relu', solver='adam', max_iter=10)\n",
    "cv = KFold(n_splits=10, random_state=1, shuffle=True)\n",
    "scores = cross_val_score(MLP, Onehot_X, Y, scoring='accuracy', cv=cv, n_jobs=-1)\n",
    "MLPMEAN = np.mean(scores)\n",
    "print(\"Accuracy:\", MLPMEAN)\n",
    "\n",
    "# hidden layer sizes: number of nodes in each layer\n",
    "# activation function: what is the fucntion that each datapoint represents?\n",
    "# solver: optimization algorithm\n",
    "# max_iter: maximum number of iterations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# REGULARIZED REGRESSION\n",
    "- Regularized regression is a type of regression is a special regression that helps to reduce the complexity of the model and prevent overfitting.\n",
    "- there are two types of regularized regression: Lasso and Ridge\n",
    "- it uses a penalty system, that helps in penalizing the model if it is too complex and hence this also helps in striking a balance between bias and variance.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9.Lasso Regression (L1 Regularization, comes with Logistic Regression)\n",
    "- lasso regression stands for Least Absolute Shrinkage and Selection Operator\n",
    "- It works well when you have many features and want to reduce the number of features used in the model.\n",
    "- feature selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.9246537396121883\n"
     ]
    }
   ],
   "source": [
    "LRL1 = LogisticRegression(penalty = 'l1', solver = 'saga', C = 0.1)\n",
    "# penalty introduced penalty system, solver is the optimization algorithm, C is the regularization parameter\n",
    "cv = KFold(n_splits=10, random_state=1, shuffle=True)\n",
    "scores = cross_val_score(LRL1, Onehot_X, Y, scoring='accuracy', cv=cv, n_jobs=-1)\n",
    "LRL1MEAN = np.mean(scores)\n",
    "print(\"Accuracy:\", LRL1MEAN)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 10. Ridge Regression (L2 Regularization, comes with Logistic Regression)\n",
    "- Ridge penalizes the sum of the squared values of the coefficients, preventing them from growing too large.\n",
    "- useful when you have a large number of co-related features and to prevent overfitting \n",
    "- does not perform feature selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.948753462603878\n"
     ]
    }
   ],
   "source": [
    "LRL2 = LogisticRegression(penalty = 'l2', solver = 'saga', C = 0.1)\n",
    "cv = KFold(n_splits=10, random_state=1, shuffle=True)\n",
    "scores = cross_val_score(LRL2, Onehot_X, Y, scoring='accuracy', cv=cv, n_jobs=-1)\n",
    "LRL2MEAN = np.mean(scores)\n",
    "print(\"Accuracy:\", LRL2MEAN)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Model Evaluation\n",
    "We will now evaluate the models based on the cross validation scores and then select the best model based on the cross validation scores."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model Performance Summary:\n",
      "SVC: 0.9147\n",
      "Logistic Regression: 0.9571\n",
      "KNN: 0.9194\n",
      "Decision Tree: 0.9593\n",
      "Random Forest: 0.9540\n",
      "Naive Bayes: 0.9122\n",
      "MLP: 0.8360\n",
      "Lasso: 0.9247\n",
      "Ridge: 0.9488\n",
      "\n",
      "Best Performing Model:\n",
      "Decision Tree with accuracy: 0.9593\n"
     ]
    }
   ],
   "source": [
    "model_scores = {\n",
    "    'SVC': SVCMEAN,\n",
    "    'Logistic Regression': LRMEAN,\n",
    "    'KNN': KNNMEAN,\n",
    "    'Decision Tree': DTMEAN,\n",
    "    'Random Forest': RTCMEAN,\n",
    "    'Naive Bayes': NBMEAN,\n",
    "    'MLP': MLPMEAN,\n",
    "    'Lasso': LRL1MEAN,\n",
    "    'Ridge': LRL2MEAN\n",
    "}\n",
    "\n",
    "best_model = max(model_scores.items(), key=lambda x: x[1])\n",
    "\n",
    "print(\"Model Performance Summary:\")\n",
    "for model, score in model_scores.items():\n",
    "    print(f\"{model}: {score:.4f}\")\n",
    "\n",
    "print(\"\\nBest Performing Model:\")\n",
    "print(f\"{best_model[0]} with accuracy: {best_model[1]:.4f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So we can infer that the best model is Random Forest Classifier with a cross validation score of 0.97\n",
    "\n",
    "Now lets train the model on the entire training data and then predict the values of the test data.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data splitting for best model\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove rows where Y contains '-', 'n', or 'y'\n",
    "Y_cleaned = Y[~Y.isin(['-', 'n', 'y'])]\n",
    "\n",
    "# Filter the corresponding rows in Onehot_X as well\n",
    "Onehot_X_cleaned = Onehot_X[Y.isin(['-', 'n', 'y']) == False]\n",
    "\n",
    "# Now split the cleaned data\n",
    "x_train, x_test, y_train, y_test = train_test_split(Onehot_X_cleaned, Y_cleaned, test_size=0.2, random_state=42)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Model prediction and evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.9230769230769231\n",
      "Recall: 0.9230769230769231\n",
      "Precision: 0.9506454772079772\n",
      "                            precision    recall  f1-score   support\n",
      "\n",
      "                         -       1.00      1.00      1.00         3\n",
      "                    cancer       1.00      1.00      1.00         7\n",
      "                 cirrhosis       1.00      1.00      1.00        21\n",
      "         ibd_crohn_disease       1.00      0.33      0.50         6\n",
      "    ibd_ulcerative_colitis       0.85      1.00      0.92        23\n",
      "impaired_glucose_tolerance       0.56      0.90      0.69        10\n",
      "             large_adenoma       1.00      1.00      1.00         4\n",
      "                   leaness       1.00      1.00      1.00        16\n",
      "                n_relative       0.93      1.00      0.97        14\n",
      "                     obese       1.00      0.33      0.50         3\n",
      "                   obesity       1.00      1.00      1.00        34\n",
      "                overweight       0.25      1.00      0.40         1\n",
      "             small_adenoma       0.80      1.00      0.89         4\n",
      "            stec2-positive       1.00      1.00      1.00        11\n",
      "                       t2d       1.00      0.82      0.90        51\n",
      "\n",
      "                  accuracy                           0.92       208\n",
      "                 macro avg       0.89      0.89      0.85       208\n",
      "              weighted avg       0.95      0.92      0.92       208\n",
      "\n"
     ]
    }
   ],
   "source": [
    "best_model = DT\n",
    "best_model.fit(x_train, y_train)\n",
    "y_prob = best_model.predict_proba(x_test)[:, 1] \n",
    "y_pred = best_model.predict(x_test)\n",
    "accuracy = accuracy_score(y_test, y_pred)\n",
    "print(\"Accuracy:\", accuracy)\n",
    "recall = recall_score(y_test, y_pred, average='weighted')\n",
    "print(\"Recall:\", recall)\n",
    "precision = precision_score(y_test, y_pred, average='weighted')\n",
    "print(\"Precision:\", precision)\n",
    "print(classification_report(y_test, y_pred))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# comparision with Metaphlan2's outcomes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now this is a multi-class classification problem and hence we will use the accuracy_score function to calculate the accuracy of the model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metric</th>\n",
       "      <th>DecisionTrees</th>\n",
       "      <th>Metaphlan2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Accuracy</td>\n",
       "      <td>91.826923</td>\n",
       "      <td>91.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>85.273648</td>\n",
       "      <td>92.577338</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Metric  DecisionTrees Metaphlan2\n",
       "0  Accuracy      91.826923      91.56\n",
       "1  f1_score      85.273648  92.577338"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data = {\n",
    "    'Metric': ['Accuracy','f1_score' ],\n",
    "    'DecisionTrees': [(accuracy_score(y_test, y_pred))*100, (f1_score(y_test, y_pred, average='macro'))*100],\n",
    "    'Metaphlan2':['91.56', 0.9257733827354075*100]\n",
    "}\n",
    "table = pd.DataFrame(data)\n",
    "display(table)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Inference\n",
    "- For our Dataset, we have used cross validation to find the best algorithm for classification of diseases.\n",
    "- We have found that Random Forest Classifier is the best algorithm for our dataset as it shows the highest cross validation score of 0.97\n",
    "- We have trained the Random Forest Classifier on the entire training data and then predicted the values of the test data.\n",
    "- We have noticed that DecisionTrees"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ROC CURVE\n",
    "\n",
    "This ROC CURVE, isn't like your normal daily ROC curves and the Reason being, there are multiple Classes and we will get multiple probabilities from the 'predict_proba' function\n",
    "\n",
    "Here we will be plotting the ROC curve for each class and then calculating the AUC for each class. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ML MODEL FOR EVERY DISEASE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from sklearn.metrics import roc_curve, auc\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Clean the data first\n",
    "mask = ~Y.isin(['-', 'n', 'y'])\n",
    "Y_clean = Y[mask]\n",
    "X_clean = X[mask]\n",
    "\n",
    "# Split the cleaned data\n",
    "x_train, x_test, y_train, y_test = train_test_split(X_clean, Y_clean, test_size=0.2, random_state=42)\n",
    "\n",
    "def plot_multiclass_roc(classifier, X_test, y_test):\n",
    "    # Get unique classes\n",
    "    classes = np.unique(y_test)\n",
    "    \n",
    "    # Initialize plot\n",
    "    plt.figure(figsize=(10, 8))\n",
    "    \n",
    "    # Calculate ROC curve for each class\n",
    "    for i, class_name in enumerate(classes):\n",
    "        # Convert to binary problem (one-vs-rest)\n",
    "        y_binary = (y_test == class_name).astype(int)\n",
    "        \n",
    "        # Get probability scores\n",
    "        y_score = classifier.predict_proba(X_test)[:, i]\n",
    "        \n",
    "        # Calculate ROC curve\n",
    "        fpr, tpr, _ = roc_curve(y_binary, y_score)\n",
    "        roc_auc = auc(fpr, tpr)\n",
    "        \n",
    "        # Plot ROC curve\n",
    "        plt.plot(fpr, tpr, label=f'{class_name} (AUC = {roc_auc:.2f})')\n",
    "    \n",
    "    # Plot diagonal line\n",
    "    plt.plot([0, 1], [0, 1], 'k--')\n",
    "    \n",
    "    # Set plot details\n",
    "    plt.xlabel('False Positive Rate')\n",
    "    plt.ylabel('True Positive Rate')\n",
    "    plt.title('Multi-class ROC Curves')\n",
    "    plt.legend(loc='lower right')\n",
    "    plt.grid(True)\n",
    "    plt.show()\n",
    "\n",
    "# Filter numerical columns and handle categorical data first\n",
    "numeric_features = [col for col, dtype in X_clean.dtypes.items() if np.issubdtype(dtype, np.number)]\n",
    "X_clean_numeric = X_clean[numeric_features]\n",
    "\n",
    "# Train the model and use the trained model for ROC curves\n",
    "best_model.fit(X_clean_numeric, Y_clean)\n",
    "plot_multiclass_roc(best_model, x_test[numeric_features], y_test)\n",
    "# E\n",
    "# model.fit(x_train, y_train)\n",
    "# plot_multiclass_roc(model, x_test, y_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.preprocessing import label_binarize\n",
    "\n",
    "def calculate_roc_scores(classifier, X_test, y_test):\n",
    "    \"\"\"\n",
    "    Calculate ROC AUC scores for multi-class classification\n",
    "    Returns dictionary with class-wise and average scores\n",
    "    \"\"\"\n",
    "    # Get probabilities and classes\n",
    "    y_prob = classifier.predict_proba(X_test)\n",
    "    classes = classifier.classes_\n",
    "    \n",
    "    # Initialize results dictionary\n",
    "    roc_scores = {}\n",
    "    \n",
    "    # Calculate ROC AUC for each class\n",
    "    for i, class_name in enumerate(classes):\n",
    "        y_true_binary = (y_test == class_name).astype(int)\n",
    "        roc_scores[f'Class_{class_name}'] = roc_auc_score(y_true_binary, y_prob[:, i])\n",
    "    \n",
    "    # Calculate micro and macro average\n",
    "    y_bin = label_binarize(y_test, classes=classes)\n",
    "    roc_scores['macro_avg'] = roc_auc_score(y_bin, y_prob, average='macro')\n",
    "    roc_scores['micro_avg'] = roc_auc_score(y_bin, y_prob, average='micro')\n",
    "    \n",
    "    return roc_scores\n",
    "\n",
    "# Usage example:\n",
    "scores = calculate_roc_scores(best_model, x_test_clean, y_test_clean)\n",
    "for metric, score in scores.items():\n",
    "    print(f\"{metric}: {score:.3f}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
